File: backend/vortex-admin-pro/src/main/java/com/vortexadmin/service/BillingService.java
package com.vortexadmin.service;

import com.vortexadmin.dto.request.UpgradePlanRequest;
import com.vortexadmin.dto.response.DiscountEligibilityResponse;
import com.vortexadmin.dto.response.InvoiceResponse;
import com.vortexadmin.dto.response.PlanResponse;
import com.vortexadmin.dto.response.SubscriptionResponse;

import java.util.List;

/**
 * Service contract for billing and subscription management, covering plan listing, subscription
 * retrieval, invoice history, plan upgrades, cancellation, and discount eligibility checks.
 */
public interface BillingService {

    /**
     * Returns all available subscription plans offered by the platform.
     *
     * @return a list of plan responses describing each available tier
     */
    List<PlanResponse> getPlans();

    /**
     * Returns the active subscription for the specified organization.
     *
     * @param organizationId the primary key of the organization
     * @return the current subscription response for the given organization
     * @throws com.vortexadmin.exception.ApiException if no active subscription is found
     */
    SubscriptionResponse getSubscription(Long organizationId);

    /**
     * Returns the invoice history for the specified organization, ordered from newest to oldest.
     *
     * @param organizationId the primary key of the organization
     * @return a list of invoice responses for the given organization
     */
    List<InvoiceResponse> getInvoices(Long organizationId);

    /**
     * Upgrades or changes the subscription plan for the organization specified in the request,
     * generates a new invoice for the prorated amount, and returns the updated subscription.
     *
     * @param request the upgrade payload containing the organization ID and the target plan ID
     * @return the updated subscription response reflecting the new plan
     * @throws com.vortexadmin.exception.ApiException if the organization or plan is not found
     */
    SubscriptionResponse upgradePlan(UpgradePlanRequest request);

    /**
     * Cancels the active subscription for the specified organization by setting its status to
     * {@code CANCELLED}.
     *
     * @param organizationId the primary key of the organization whose subscription should be cancelled
     * @return the updated subscription response with {@code CANCELLED} status
     * @throws com.vortexadmin.exception.ApiException if no active subscription is found
     */
    SubscriptionResponse cancelSubscription(Long organizationId);

    /**
     * Checks whether the specified organization qualifies for a discount based on criteria
     * such as subscription duration or member count.
     *
     * @param organizationId the primary key of the organization to evaluate
     * @return a {@link DiscountEligibilityResponse} indicating eligibility and the applicable
     *         discount percentage if eligible
     */
    DiscountEligibilityResponse getDiscountEligibility(Long organizationId);
}

File: frontend/src/services/billingService.js
import api from '../api/axios';

/**
 * Service object for managing billing plans, subscriptions, invoices, and
 * discounts for organizations.
 */
export const billingService = {
    /**
     * Fetches all available billing plans.
     * GET /billing/plans
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    getPlans: () => api.get('/billing/plans'),

    /**
     * Fetches the current subscription for a specific organization.
     * GET /billing/subscription?organizationId={organizationId}
     * @param {number|string} organizationId - The organization whose subscription to retrieve.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    getSubscription: (organizationId) => api.get('/billing/subscription', { params: { organizationId } }),

    /**
     * Fetches invoice history for a specific organization.
     * GET /billing/invoices?organizationId={organizationId}
     * @param {number|string} organizationId - The organization whose invoices to retrieve.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    getInvoices: (organizationId) => api.get('/billing/invoices', { params: { organizationId } }),

    /**
     * Upgrades an organization's subscription to the specified plan.
     * POST /billing/upgrade
     * @param {number|string} organizationId - Target organization ID.
     * @param {string} planName - Plan identifier (e.g. 'PRO', 'ENTERPRISE').
     * @param {string} [billingCycle='MONTHLY'] - Billing interval: 'MONTHLY' | 'YEARLY'.
     * @param {string} [paymentProvider='MOCK'] - Payment gateway to use.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    upgrade: (organizationId, planName, billingCycle = 'MONTHLY', paymentProvider = 'MOCK') =>
        api.post('/billing/upgrade', { organizationId, planName, billingCycle, paymentProvider }),

    /**
     * Cancels an organization's active subscription.
     * POST /billing/cancel?organizationId={organizationId}
     * @param {number|string} organizationId - The organization whose subscription to cancel.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    cancel: (organizationId) => api.post('/billing/cancel', null, { params: { organizationId } }),

    /**
     * Fetches active discount codes or promotions for a specific organization.
     * GET /billing/discounts?organizationId={organizationId}
     * @param {number|string} organizationId - The organization to query discounts for.
     * @returns {Promise<import('axios').AxiosResponse>}
     */
    getDiscounts: (organizationId) => api.get('/billing/discounts', { params: { organizationId } }),
};
